content-src/components/DiscoveryStreamBase/DiscoveryStreamBase.jsx

```diff
@@ -169,10 +169,7 @@ export class _DiscoveryStreamBase extends React.PureComponent {
   render() {
     // Select layout render data by adding spocs and position to recommendations
     const layoutRender = selectLayoutRender(this.props.DiscoveryStream, this.props.Prefs.values, rickRollCache);
-    const {config, feeds, spocs} = this.props.DiscoveryStream;
-    if (!spocs.loaded || !feeds.loaded) {
-      return null;
-    }
```

Comment:
With this removal we are no longer waiting for both spocs and feeds to finish loading before rendering. Are we checking later for a component we have both spocs and feeds before rendering?

Reply:
Yup, this block right here in selectLayoutRender is the key:

```
      for (const component of row.components.filter(c => !filterArray.includes(c.type))) {
        if (component.feed) {
          const spocsConfig = component.spocs;
          // Are we still waiting on a feed/spocs, render what we have, and bail out early.
          if (!feeds.data[component.feed.url] ||
            (spocsConfig && spocsConfig.positions && spocsConfig.positions.length && !spocs.loaded)) {
            return renderedLayoutArray;
          }
          components.push(handleComponent(component));
        } else {
          components.push(component);
        }
      }
```

First it's if it's not a feed based component, if it expects a feed, then it's checking if the feed exists yet, if it doesn't it bails and returns what it has up to this point. So if feed 1 is done, and feed 2 is loading still, we can render feed 1 and keep waiting on feed 2. It the checks to see if this feed expects a spoc and if spocs are loaded, if not, it considers the whole thing not ready too, and returns what we have, and keeps waiting.

Reply:
So it's less of a "check to make sure we have everything before rendering" and more of a "filter out the stuff we cannot render before rendering"

Reply:
Since spocs.loaded is a needed for rendering if a feed expects a spoc , should we have loadSpocs before loadComponentFeeds in refreshAll https://github.com/mozilla/activity-stream/pull/4909/files#diff-8b1b03c4d5dec5c8e4f7de50e05c0c4fR679